Concurrent reads

kderme opened the pull request:
This pr keeps a single read handle per file of the VolatileDB, so that it reduces reopening for each read https://github.com/input-output-hk/ouroboros-network/issues/767. A new read handle opens when we parse a file of the db (during opening the db) or when a new file is created. 

It is important that these files are properly garbage collected, which is not trivial.
q-s-m tests caught some cases where open handles leak, which I fixed with exception-handling, but a more all-around solution should be given as part of https://github.com/input-output-hk/ouroboros-network/issues/649, maybe using `ResourceRegistry`.

`pread` is used to allow concurrent readers. `pread` (in unix) is given a read lock on the FS layer.
Also ImmutableDB now uses pread in some cases.

~First prs are also part of https://github.com/input-output-hk/ouroboros-network/pull/1160 and https://github.com/input-output-hk/ouroboros-network/pull/1164~

Last piece of https://github.com/input-output-hk/ouroboros-network/pull/1149

mrBliss commented:
Can you make a separate PR for the changes to the ImmutableDB? Just using `pread` instead of `seek` + `read` doesn't give us any more concurrency until we let readers concurrently access the state.

kderme commented:
> Can you make a separate PR for the changes to the ImmutableDB? Just using `pread` instead of `seek` + `read` doesn't give us any more concurrency until we let readers concurrently access the state.

Sure yes, I didn't try to increase parallelism in the ImmutableDB, I just added it based on this comment https://github.com/input-output-hk/ouroboros-network/issues/1140#issuecomment-542648821
Do you want me also to create another issue for parallelism in ImmutableDB?

kderme commented on ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/FileInfo.hs:
```diff
@@ -59,35 +63,38 @@ data FileSlotInfo blockId = FileSlotInfo {
-addSlot slotNo slotOffset slotInfo MkFileInfo{..} = MkFileInfo {
+        -> FileInfo blockId h
+        -> FileInfo blockId h
+addSlot slotNo slotOffset slotInfo fileInfo@MkFileInfo{..} = fileInfo {
```
Right, I changed it since `fReadHandle` stays the same, but forgot to remove it.

kderme replied:
Note that I no longer give all the fields, `fReadHandle` stays the same.

kderme commented on ouroboros-consensus/src/Ouroboros/Storage/FS/Handle.hs:
```diff
@@ -30,36 +33,48 @@ instance Eq (HandleOS a) where
+-- Multiple readers are allowed. Writers block if there are any readers. Readers
+-- block if there is any writer.
+readOpenHandle :: String -> HandleOS osHandle -> (osHandle -> IO a) -> IO a
+readOpenHandle label (HandleOS fp hVar) k =
```
I like this better

> withSharedOpenHandle and withExclOpenHandle

kderme commented on ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs:
```diff
@@ -633,25 +659,33 @@ modifyState VolatileDBEnv{_dbHasFS = hasFS :: HasFS m h, ..} action = do
+    closeWithState VolatileDbClosed = return ()
+    closeWithState (VolatileDbOpen InternalState {..}) = do
+      hClose _currentWriteHandle
+      mapM_ (hClose . FileInfo.getHandle) $ Index.elems _currentMap
```
> What if closing the first handle fails?
> 
> I think it would be good to use a `ResourceRegistry`for the handles. When closing the `VolatileDB`, you simply close the registry, which is already [able](https://github.com/input-output-hk/ouroboros-network/blob/e9430d56c1378ff31d05ddd63b7f964a6ff2c293/ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs#L539) to deal with exceptions thrown in the clean-up actions.

I have some questions about using `ResourceRegistry` for this:
- Will it be the same registry as the `cdbRegistry` of `ChainDbArgs` or a new one should be created for the VolatileDB? I guess the latter or else closing the registry from the VolatileDB would also release resources of the ChainDB.
- Even if VolatileDB has its own dedicated `ResourceRegistry`, it's possible to close and reopen the db, but the registry can't reopen. Also `closeRegistry` is only exported for tests and has to be used by the same thread that opened it. However in the VolatileDB I can't make sure it's the same thread. Maybe a new function `relaseAll` or `cleanRegistry` should be added for this.
- `ResourceKey` is not exported from `Ouroboros.Consensus.Util.ResourceRegistry`. Is there any reason for this or should I just export it? (it's used to call release).

mrBliss replied:
> > What if closing the first handle fails?
> > I think it would be good to use a `ResourceRegistry`for the handles. When closing the `VolatileDB`, you simply close the registry, which is already [able](https://github.com/input-output-hk/ouroboros-network/blob/e9430d56c1378ff31d05ddd63b7f964a6ff2c293/ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs#L539) to deal with exceptions thrown in the clean-up actions.
> 
> I have some questions about using `ResourceRegistry` for this:
> 
>     * Will it be the same registry as the `cdbRegistry` of `ChainDbArgs` or a new one should be created for the VolatileDB? I guess the latter or else closing the registry from the VolatileDB would also release resources of the ChainDB.

Indeed a new one for the VolatileDB.

> 
>     * Even if VolatileDB has its own dedicated `ResourceRegistry`, it's possible to close and reopen the db, but the registry can't reopen. Also `closeRegistry` is only exported for tests and has to be used by the same thread that opened it. However in the VolatileDB I can't make sure it's the same thread. Maybe a new function `relaseAll` or `cleanRegistry` should be added for this.

Hmm, can't you create a new registry when reopening? Good point about it needing to be the same thread :thinking: 
> 
>     * `ResourceKey` is not exported from `Ouroboros.Consensus.Util.ResourceRegistry`. Is there any reason for this or should I just export it (opaque)?

Yeah, I think it was just forgotten.

kderme replied:
> Hmm, can't you create a new registry when reopening?

Maybe it is possible, but in this case I have to use the `openRegistry`/`closeRegistry` combinators instead of `withRegistry`. I think the VolatileDB should not create and close the registry: it should be given to it as an argument when opening, like in mempool. The registry should be created when the node starts like in `Ouroboros.Consensus.Node`. That's why I was thinking about a `releaseAll` function. 

>  Good point about it needing to be the same thread :thinking:

A `releaseAllunsafe` function which does not check about threads would also fix this. There are already other locks in the VolatileDB which protect against, for example, trying to use a handles which is already closed.

kderme commented:
> I think there is a potential race condition, see my comment.

Which race condition/comment are you referring to? There was a race condition, because I missed a needed putMVar in `withState`, is this what you mean?

kderme commented:
I added a new flag for the creation of a `ResourceRegistry`. When this flag is not set, no checks are performed related to which threads use the registry. 

Also some tests failed. This is now fixed.

mrBliss commented on ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs:
```diff
@@ -211,25 +216,25 @@ openDBImpl :: (HasCallStack, IOLike m, Ord blockId, NoUnexpectedThunks blockId)
+closeDBImpl VolatileDBEnv{..} =
+    bracket (takeMVar _dbInternalState)
+            (\_ -> putMVar _dbInternalState VolatileDbClosed)
+            (\_ -> unsafeReleaseAll _registry)
```
Note the `wrapFsError` that we had, where does it live now? It wrapped `FsErrror`s in `UnexpectedError.FileSystemError`.

mrBliss commented on ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs:
```diff
@@ -620,42 +694,38 @@ modifyState VolatileDBEnv{_dbHasFS = hasFS :: HasFS m h, ..} action = do
-        closeOpenHandle mst
+      ExitCaseException _ex ->
+        unsafeReleaseAll _registry
+          `finally` putMVar _dbInternalState VolatileDbClosed
```
Why not:
```haskell
        putMVar _dbInternalState VolatileDbClosed
        unsafeReleaseAll _registry
```
The `putMVar` should never fail, right?

kderme replied:
I think it's the same case as this https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345820664
cleaning the registry should happen with the `InternalState` lock.

mrBliss commented on ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs:
```diff
@@ -457,34 +472,46 @@ nextFile HasFS{..} _err VolatileDBEnv{..} st@InternalState{..} = do
+    -- In this way, any pending read operation will fail with a
+    -- @read: FHandle closed@ user error. If we don't close the handle
+    -- a reader may read from a offset bigger than the size of the file,
+    -- which we shouldn't allow.
```
```
In this way, any pending read operation will fail with a
@read: FHandle closed@ user error.
```

Hmm, that's not ideal. We should think about this.

mrBliss commented:
> > I think there is a potential race condition, see my comment.
> 
> Which race condition/comment are you referring to? There was a race condition, because I missed a needed putMVar in `withState`, is this what you mean?

Huh, I can't seem to find it either. It was a huge comment with a table and all, aargh, GitHub :frowning_face:. I'll search for it or write it again.

mrBliss commented on ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs:
```diff
@@ -357,17 +367,21 @@ garbageCollectImpl env@VolatileDBEnv{..} slot =
-            removeFile $ filePath fileId
+            let file = filePath fileId
+            -- First close the read handle.
+            forM_ (Index.getHandle fileId _currentMap) releaseHandle
```
I had written an extensive comment with tables and all about a race condition here, but GitHub lost it :cry:. A simpler rewrite:

There's a race condition here. Consider the following sequence of events:

1. Thread A calls `getBlock`, the state is read in `withState` using `readMVar`. Note that the state is still left in the `MVar`. The file to read the block from is file X. Before thread A can read the block using the handle for file X, the next thing happens:
2. Thread B calls `garbageCollect`. It obtains the same state from the `MVar` as thread A did. It decides to garbage collect file X, so it releases/closes its handle.
3. Thread A tries to read its block from the handle for file X that thread B just closed. Error: handle closed :boom:

Note that the closing a handle that is being used blocks until it is no longer used, but that doesn't help here, because we'll use the handle after it has been closed.

To be entirely safe  `getBlock` and `garbageCollect` should not happen concurrently. This could be enforced by something akin to `ReadWriteVar`, but we would need to roll our own version as we're using the IOSim.

Moreover, `getBlock` and `putBlock` *can* happen concurrently. `putBlock` and `garbageCollect` can *almost* happen concurrently, as it is very unlikely that we need to garbage collect the file we're appending to. Maybe it is even worth to make this impossible?

kderme replied:
This is similar to what I try to explain here https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345673029.
But I guess this situation is expected if we want concurrent reads: If the user tries to get a block, while another thread makes this block obsolete by garbage collection, this should be an error. I also tried to control this situation: by using exclusive locks in the fs layer, read(2) and close(2) can't happen concurrently, which is an undefined behavior.

Maybe it's worth improving the api in these case: for example wrapping this exception with a better `VolatileDBError` or just return Nothing. I like the latter a bit more, since getBlock returns Nothing when a block is not found, however it has the disadvantage that it may hide other `handle closed` errors for other reasons. In my opinion, after a `handle closed` error, the best solution is to read the InternalState MVar again,  and see if the file is really garbage collected and only in this case return Nothing.

> putBlock and garbageCollect can almost happen concurrently, as it is very unlikely that we need to garbage collect the file we're appending to

This shouldn't be an issue, since these operations are serialized by the `InternalState` MVar

mrBliss replied:
> This is similar to what I try to explain here [#1165 (comment)](https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345673029).
> But I guess this situation is expected if we want concurrent reads: If the user tries to get a block, while another thread makes this block obsolete by garbage collection, this should be an error. I also tried to control this situation: by using exclusive locks in the fs layer, read(2) and close(2) can't happen concurrently, which is an undefined behavior.

Should it be an error, though? Returning `Nothing`, as you say in the next paragraph, seems like a reasonable thing to do.

> > This could be enforced by something akin to ReadWriteVar, but we would need to roll our own version as we're using the IOSim.
> 
> I'm afraid adding another lock would complicate things a bit too much.

True.

> 
> Maybe it's worth improving the api in these case: for example wrapping this exception with a better `VolatileDBError` or just return Nothing. I like the latter a bit more, since getBlock returns Nothing when a block is not found, however it has the disadvantage that it may hide other `handle closed` errors for other reasons. In my opinion, after a `handle closed` error, the best solution is to read the InternalState MVar again, and see if the file is really garbage collected and only in this case return Nothing.

That sounds good to me. If you implement this, add a comment to `getBlockImpl` saying this should not loop because a handle is only closed when the file is garbage collected (so we will find `Nothing` in the recursive call) or when the VolatileDB itself is closed (so we will get a `ClosedDBError` in the recursive call). (Grep for `errLoad` to see an example of where I only catch a certain error.)

> > putBlock and garbageCollect can almost happen concurrently, as it is very unlikely that we need to garbage collect the file we're appending to
> 
> This shouldn't be an issue, since these operations are serialized by the `InternalState` MVar

Yes, but my point is actually the opposite: we could allow them to happen concurrently if we make it impossible to garbage-collect the current file. Then garbage collection wouldn't block adding new blocks, which would speed up syncing, as adding a block to the VolatileDB is a synchronous step in the syncing process.

However, I'm inclined to think garbage collection can happen so fast that we don't need to bother with this (added complexity). Maybe if we measure this in the future with some benchmarks and we see a GC causing a noticeable drop in sync speed, we can think about it again.

Nonetheless, I think the current file should be never be garbage collected. In all properly configured VolatileDBs, this should not never be needed anyway. Getting rid of this special case (+ `reOpenFile`) will simplify things. Do you agree?

kderme commented on ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs:
```diff
@@ -179,20 +182,22 @@ openDB :: (HasCallStack, IOLike m, Ord blockId, NoUnexpectedThunks blockId)
        -> ThrowCantCatch (VolatileDBError blockId) (STM m)
        -> Parser e m blockId
        -> Int
+       -> ResourceRegistry m
```
This is mainly because of how VolatileDB is initialized. There is nothing that makes sure that closeDB is called, since `withDB` is not used, but `openDB` and `closeDB`. By creating a registry and passing it as an argument, this issue is fixed. I also think this approach fits well how the code already handles other resources.

kderme commented on ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs:
```diff
@@ -211,25 +216,25 @@ openDBImpl :: (HasCallStack, IOLike m, Ord blockId, NoUnexpectedThunks blockId)
+closeDBImpl VolatileDBEnv{..} =
+    bracket (takeMVar _dbInternalState)
+            (\_ -> putMVar _dbInternalState VolatileDbClosed)
+            (\_ -> unsafeReleaseAll _registry)
```
I wasn't very sure about how thread safe `unsafeReleaseAll` is. This is why I always call it with the InternalState lock.
But I think you're right, this is not necessary.

kderme replied:
Hmm now that I think about it again, it may be possible that after swapping the Mvar to VolatileDBClosed and before cleaning the registry, another thread may reopen the db and allocate new resources, which should not be cleaned.

Before the registry, this was not possible, since we only closed the "old" handles, but with the registry we have scope to the new handles.

mrBliss replied:
> Hmm now that I think about it again, it's possible that after swapping the Mvar to VolatileDBClosed and before cleaning the registry, another thread may reopen the db and allocate new resources, which should not be cleaned.

Although it won't happen in practice, as we never reopen a VolatileDB (but we do in the tests, right?), you are technically correct. So you can leave the `bracket`.

kderme commented:
Adrressed:
https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345665620 
https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345659628
https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345634853
https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345634764

I think this was already fixed https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345640747

mrBliss commented:
> Addressed:
> [#1165 (comment)](https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345665620)
> [#1165 (comment)](https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345659628)
> [#1165 (comment)](https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345634853)
> [#1165 (comment)](https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345634764)
> 
> I think this was already fixed [#1165 (comment)](https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345640747)

It's easier if you reply to each comment separately, and if it is just a simple typo fix, you can mark it resolved yourself.

mrBliss commented on ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:
```diff
@@ -1313,11 +1316,20 @@ mkArgs :: IOLike m
        -> ChainDbArgs m Blk
-mkArgs cfg initLedger tracer registry varCurSlot
+mkArgs cfg initLedger tracer registry registryVolDB
+>>>>>>> Use ResourceRegistry for the handles in VolatileDB
```
Committed merge conflict. There are two more in this file.

mrBliss commented on ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/AddBlock.hs:
```diff
@@ -114,8 +114,14 @@ prop_addBlock_multiple_threads bpt =
           <$> uncheckedNewTVarM Mock.empty
           <*> uncheckedNewTVarM Mock.empty
           <*> uncheckedNewTVarM Mock.empty
+<<<<<<< 70c4adec8c3262270d7b4fb9aca42fc04e2a5a66
```
There are committed merge conflicts in this file.

kderme commented:
> I thought using the `ResourceRegistry` would help us make sure the open handles are closed in every scenario at a low (complexity) cost. I thought it would be something entirely internal to the VolatileDB (you haven't convinced me of the opposite, though). Releasing from a different thread was an unforeseen issue.

Do you think the complexity mainly comes from having to close form a different thread (and all the unsafe Registry stuff that I had to introduce) or from moving around an additional registryVolDB? If it's the latter, I think I have an idea to make it internal to the VolatileDB: instead of nesting registries:
`withRegistry \registry -> withRegistry \registryVolDB -> do ...`

I can actually do
```
withRegistry \registry -> do
    ...
   openVolDB registry ...
    ...

openVolDB registry ... = do
   ...
   allocate registry (\ _ -> newUnsafeRegistry) closeRegistry
```
In that way the registry is seen just as another resource and will always be closed when the node does its cleanup. Opening the volatileDB will require to pass the registry, but the registryVolDB will remain internal.

Do you think this is enough to keep the registry logic? Or do you find that the unsafe Registry stuff is by itself too complicated to bother?

mrBliss commented:
> > I thought using the `ResourceRegistry` would help us make sure the open handles are closed in every scenario at a low (complexity) cost. I thought it would be something entirely internal to the VolatileDB (you haven't convinced me of the opposite, though). Releasing from a different thread was an unforeseen issue.
> 
> Do you think the complexity mainly comes from having to close form a different thread (and all the unsafe Registry stuff that I had to introduce) or from moving around an additional registryVolDB? If it's the latter, I think I have an idea to make it internal to the VolatileDB: instead of nesting registries:
> `withRegistry \registry -> withRegistry \registryVolDB -> do ...`
> 
> I can actually do
> 
> ```
> withRegistry \registry -> do
>     ...
>    openVolDB registry ...
>     ...
> 
> openVolDB registry ... = do
>    ...
>    registryVolDB <- allocate registry (\ _ -> newUnsafeRegistry) closeRegistry
> ```
> 
> In that way the registry is seen just as another resource and will always be closed when the node does its cleanup. Opening the volatileDB will require to pass the registry, but the registryVolDB will remain internal.
> 
> Do you think this is enough to keep the registry logic? Or do you find that the unsafe Registry stuff is by itself too complicated to bother?

The fact that it has "unsafe" in its name indicates to me that is the wrong solution.

kderme commented:
> The fact that it has "unsafe" in its name indicates to me that is the wrong solution.

Unsafe means here that any thread can use it. The volatileDB makes it safe, in the sense that it is protected by the internalState lock.

In any way I agree, the handles will be closed when the process dies, so the complexity here may be unnecessary. I'll undo the changes. Just something to keep in mind: some scripts may reopen the node and they should try to let os do any cleanup before actually reopening.

mrBliss commented:
> > The fact that it has "unsafe" in its name indicates to me that is the wrong solution.
> 
> Unsafe means here that any thread can use it. The volatileDB makes it safe, in the sense that it is protected by the internalState lock.

Right, but I still think it's not worth the additional complexity (including `HandleResource`).

> In any way I agree, the handles will be closed when the process dies, so the complexity here may be unnecessary. I'll undo the changes. Just something to keep in mind: some scripts may reopen the node and they should try to let os do any cleanup before actually reopening.
A restart of the node is: shut down the whole process, including the ChainDB and thus VolatileDB, at which point the OS releases the remaining file handles, then start the process again. So this is taken care of.
